index.js

```diff
@@ -1128,90 +703,183 @@
+    return t.type === 'UNARY' || t.type === 'BINARY';
   };
 
-  //  showValuesAndTypes :: (ErrorRepr, String) -> String
-  var showValuesAndTypes = function(repr, pad) {
-    return repr.isParameterizedType ?
-      showParameterizedValuesAndTypes(repr.type, repr.values, pad) :
-      showValueAndType(repr.types, repr.values);
+  //  showType :: Type -> String
+  var showType = function(t) {
+    var s = String(t);
+    return isParameterizedType(t) ? s.slice(1, -1) : s;
   };
 
-  //  showErrorValue :: (Integer, ErrorRepr) -> String
-  var showErrorValue = function(num, repr) {
-    var numRepr = num + ')  ';
-    return (
-      numRepr +
-      showValuesAndTypes(repr, numRepr.length) +
-      '\n\n'
+  //  showTypeSig :: [Type] -> String
+  var showTypeSig = function(types) {
+    return arrowJoin(map(types, showType));
+  };
+
+  //  showTypeSig_ :: [Type] -> String
+  var showTypeSig_ = function(types) {
+    return arrowJoin(map(types, showType).concat(['']));
+  };
+
+  //  _showTypeSig :: [Type] -> String
+  var _showTypeSig = function(types) {
+    return arrowJoin([''].concat(map(types, showType)));
+  };
+
+  //  _showTypeSig_ :: [Type] -> String
+  var _showTypeSig_ = function(types) {
+    return arrowJoin([''].concat(map(types, showType)).concat(['']));
+  };
+
+  //  showValueAndType :: Pair Any [Type] -> String
+  var showValueAndType = function(pair) {
+    return show(pair[0]) + ' :: ' + map(pair[1], showType).join(', ');
+  };
+
+  //  constraintViolation ::
+  //  (String, StrMap [Type], [Type], Any, Integer, [Type], String, TypeClass)
+  //  -> Error
+  var constraintViolation = function(name, constraints, expTypes, value, index,
+                                     actualTypes, typeVarName, typeClass) {
+    var reprs = chain(toPairs(constraints), function(pair) {
+      return map(pair[1], function(tc) {
+        var match = tc.name === typeClass.name && pair[0] === typeVarName;
+        return r(match ? '^' : ' ')(stripNamespace(tc.name) + ' ' + pair[0]);
+      });
+    });
+
+    var carets = when(reprs.length > 1,
+                      function(s) { return _('(') + s + _(')'); },
+                      reprs.join(_(', ')));
+
+    var padding = _(showTypeSig_(expTypes.slice(0, index)));
+
+    return new TypeError(
+      'Type-class constraint violation\n\n' +
+      name + ' :: ' + constraintsRepr(constraints) +
+      showTypeSig(expTypes) + '\n' +
+      _(name + ' :: ') + carets + _(' => ') +
+      padding + r('^')(typeVarName) + '\n' +
+      _(name + ' :: ' + carets + ' => ') +
+      padding + label('1')(typeVarName) + '\n\n' +
+      '1)  ' + showValueAndType([value, actualTypes]) + '\n\n' +
+      LEFT_SINGLE_QUOTATION_MARK + name + RIGHT_SINGLE_QUOTATION_MARK +
+      ' requires ' +
+      LEFT_SINGLE_QUOTATION_MARK + typeVarName + RIGHT_SINGLE_QUOTATION_MARK +
+      ' to satisfy the ' + typeClass + ' type-class constraint;' +
+      ' the value at position 1 does not.\n'
```

Comment:
What do you think about adding

`// jscs:disable maximumLineLength`

to the multi-line error message functions so we could write something like

```js
return new TypeError(
  'Type-class constraint violation\n' +
  '\n' +
  name + ' :: ' + constraintsRepr(constraints) + showTypeSig(expTypes) + '\n' +
  _(name + ' :: ') + carets + _(' => ') + padding + r('^')(typeVarName) + '\n' +
  _(name + ' :: ' + carets + ' => ') + padding + label('1')(typeVarName) + '\n' +
  '\n' +
...
```
For me this is a lot easier to read/understand, similar to how we lined up the error messages in `index/test.js`